python/src/spark_rapids_ml/knn.py

```diff
@@ -686,3 +695,390 @@ def read(cls) -> MLReader:
+        )
+        k = self.getK()
+
+        query_id_col_name = f"query_{self.getIdCol()}"
+        knn_df_agg = self.__class__._agg_topk(
+            knn_df, query_id_col_name, "indices", "distances", k
+        )
+
+        return (self._item_df_withid, query_df_withid, knn_df_agg)
+
+    def _transform(self, dataset: DataFrame) -> DataFrame:
+        raise NotImplementedError(
+            "ApproximateNearestNeighborsModel does not provide a transform function. Use 'kneighbors' instead."
+        )
+
+    def _get_cuml_transform_func(
+        self, dataset: DataFrame, eval_metric_info: Optional[EvalMetricInfo] = None
+    ) -> Tuple[
+        _ConstructFunc,
+        _TransformFunc,
+        Optional[_EvaluateFunc],
+    ]:
+
+        cuml_alg_params = self.cuml_params.copy()
+
+        def _construct_sgnn() -> CumlT:
+
+            from cuml.neighbors import NearestNeighbors as SGNN
+
+            nn_object = SGNN(output_type="cupy", **cuml_alg_params)
+
+            return nn_object
+
+        row_number_col = alias.row_number
+        input_col, input_cols = self._get_input_columns()
+        assert input_col is not None or input_cols is not None
+        id_col_name = self.getIdCol()
+
+        bcast_qids = self.bcast_qids
+        bcast_qfeatures = self.bcast_qfeatures
+
+        assert bcast_qids is not None and bcast_qfeatures is not None
+
+        def _transform_internal(
+            nn_object: CumlT, df: Union[pd.DataFrame, np.ndarray]
+        ) -> pd.DataFrame:
+
+            item_row_number = df[row_number_col].to_numpy()
+            item = df.drop(row_number_col, axis=1)  # type: ignore
+            if input_col is not None:
+                assert len(item.columns) == 1
+                item = np.array(list(item[item.columns[0]]), order="C")
+
+            if len(item) == 0:
+                return pd.DataFrame(
+                    {
+                        "query_{id_col_name}": [],
+                        "indices": [],
+                        "distances": [],
+                    }
+                )
+
+            nn_object.fit(item)
+            import cupy as cp
+
+            distances, indices = nn_object.kneighbors(bcast_qfeatures.value)
+
+            if cuml_alg_params["algorithm"] == "ivfflat":
+                distances = distances * distances
+
+            indices = indices.get()
+            indices_global = item_row_number[indices]
+
+            res = pd.DataFrame(
+                {
+                    f"query_{id_col_name}": bcast_qids.value,
+                    "indices": list(indices_global),
+                    "distances": list(distances.get()),
+                }
+            )
+            return res
+
+        return _construct_sgnn, _transform_internal, None
+
+    def approxSimilarityJoin(
+        self,
+        query_df: DataFrame,
+        distCol: str = "distCol",
+    ) -> DataFrame:
+        """
+        TODO: merge code with NearestNeighborsModel class,
```

Comment:
The TODO and comment shouldn't be in a doc string.  It will show up in api docs.
Also would be good to have an actual docstring for this.    For api docs.

Reply:
Removed the TODO. Working on it.